Refs #35704 -- Used copy.replace() in Operation.reduce() methods.

adamchainz opened the pull request:
#### Trac ticket number

ticket-35704

#### Branch description

Two commits:
1. Resolve the reported issue with extra tests that required a little restructuring.
2. Use the same `__class__` pattern everywhere.

Note that if we used `copy.replace()` per [this comment](https://github.com/django/django/pull/18508#issuecomment-2305361645), we could also remove these `__class__` references, as it copies class implicitly.

#### Checklist
- [x] This PR targets the `main` branch. <!-- Backports will be evaluated and done by mergers, when necessary. -->
- [x] The commit message is written in past tense, mentions the ticket number, and ends with a period.
- [x] I have checked the "Has patch" ticket flag in the Trac system.
- [x] I have added or updated relevant tests.
- [x] I have added or updated relevant docs, including release notes if applicable.
- [x] I have attached screenshots in both light and dark modes for any UI changes.

sarahboyce reviewed (commented):
Thank you @adamchainz :star: 
I have tested and will merge 4046558e7c50471ff9582f04a5d5f5fba37223c4 (approved)
----

It would be nice to have regression tests for 04fef299cca39f52455c8ea0a5a9115f3c45a337

sarahboyce commented:
> ...will merge [4046558](https://github.com/django/django/commit/4046558e7c50471ff9582f04a5d5f5fba37223c4) (approved)

Merged in f5ddd54986172c29c32f5d835584ea237d5a3781

adamchainz commented:
> It would be nice to have regression tests for [04fef29](https://github.com/django/django/commit/04fef299cca39f52455c8ea0a5a9115f3c45a337)

Sure, but that’s quite a lot of work.

Let me play with a `copy.replace()` backport first.

adamchainz commented:
Okay, I rebased and added a commit using `copy.replace()`. I think it’s quite neat, despite a little awkwardness defining a generic `__replace__` method due to support for positional args.

felixxm reviewed (approved):
@adamchainz Nice cleanup :+1: I squashed commits.

felixxm commented:
I updated one more case in `AddConstraint.reduce()`.

adamchainz commented:
Thanks!
